### PR TITLE
FEATURE: Chat message bookmarks

### PR DESCRIPTION
**Pull request by martin-brennan**
**Note: Do not merge until the complete polymorphic bookmark move has been made in core.**

Registers the `ChatMessage` bookmarkable using the core API and introduces the UI to bookmark individual chat messages, along with reminder notification functionality for chat message bookmarks.

**Also Note: The design of this is not complete yet, but the rest of the PR is still fine to review!**. I will add UI/JS tests once the design is nailed down as things may move around.

Chat bookmarks in user list:

image

Bookmarks within the channel:

image
image

### CODE REVIEW

**Comment by jjaffeux** [comment]
Just saying we can merge with incomplete design and iterate later, dont block yourself on this martin.

File: lib/chat_channel_fetcher.rb
Comment (jjaffeux) [inline lib/chat_channel_fetcher.rb:19]: This should have tests I think

File: assets/javascripts/discourse/templates/components/chat-message.hbs
Comment (jjaffeux) [inline assets/javascripts/discourse/templates/components/chat-message.hbs:117]: I imagine we have the same kind of logic in core, this could be a {{bookmark-icon bookmark=bookmark}} component, which would also be easy to test

**Review by jjaffeux — APPROVED** [review]

File: assets/javascripts/discourse/components/chat-message.js
Comment (jjaffeux) [inline assets/javascripts/discourse/components/chat-message.js:807]: it's possible you currently can't do differently due to chat design, but this bookmark and message.bookmark at the same time seems like a smell to me.

**Review by jjaffeux — APPROVED** [review]

**Comment by jjaffeux** [comment]
Overall looks super good 👏👏👏

File: assets/javascripts/discourse/components/chat-message.js
Comment (martin-brennan) [reply assets/javascripts/discourse/components/chat-message.js:807]: I just did it this way because I didn't want to override `message.bookmark` which comes down from the serializer, but thinking about it more I can probably just do that without problems :+1:

File: assets/javascripts/discourse/templates/components/chat-message.hbs
Comment (martin-brennan) [reply assets/javascripts/discourse/templates/components/chat-message.hbs:117]: Actually in core all of the icons and this kind of logic for bookmarks are set in widgets. However this is a good idea, I will come back to it after merging the polymorphic bookmarks in core.

File: lib/chat_channel_fetcher.rb
Comment (martin-brennan) [reply lib/chat_channel_fetcher.rb:19]: It is tested in a way via the bookmarkable query tests, but yes I will add dedicated tests for this section because it is important.

**Comment by martin-brennan** [comment]
> Just saying we can merge with incomplete design and iterate later, dont block yourself on this martin.

Thanks @jjaffeux I will make your changes and I will hopefully merge this week after https://github.com/discourse/discourse/pull/16729 in core

File: lib/chat_channel_fetcher.rb
Comment (martin-brennan) [reply lib/chat_channel_fetcher.rb:19]: Just checked and this channel fetcher is pretty badly tested :( or maybe mostly indirectly tested. Will add a todo to myself to add more.

Comment (jjaffeux) [reply lib/chat_channel_fetcher.rb:19]: Yes fetcher is very low on tests, don't need to go aggressive on this, but at least when we add a new method it should be tested directly and not indirectly via another test.

**Review by jjaffeux — APPROVED** [review]

**Pull request merged**
